symfony/plugins/orangehrmAttendancePlugin/Api/AttendanceRecordAPI.php

```diff
@@ -0,0 +1,339 @@
+        return new EndpointResourceResult(AttendanceRecordModel::class, $attendanceRecord);
+    }
+
+    /**
+     * @param  AttendanceRecord  $attendanceRecord
+     * @return bool
+     */
+    private function isAuthUserAllowedToPerformTheActions(AttendanceRecord $attendanceRecord): bool
+    {
+        $attendanceRecordOwnedEmpNumber = $attendanceRecord->getEmployee()->getEmpNumber();
+        $loggedInUserEmpNumber = $this->getAuthUser()->getEmpNumber();
+        $rolesToInclude = [];
+        //check the configuration as ESS since Admin user is always allowed to update
+        if ($attendanceRecordOwnedEmpNumber === $loggedInUserEmpNumber) {
+            $rolesToInclude = ['ESS'];
+        }
+        //If edit own attendance record, get the allowed actions list as an ESS user
+        //If edit for someone, get the allowed actions list as a Supervisor
+        //Admin is always allowed to edit others records
+        $allowedWorkflowItems = $this->getUserRoleManager()->getAllowedActions(
+            WorkflowStateMachine::FLOW_ATTENDANCE,
+            $attendanceRecord->getState(),
+            [],
+            $rolesToInclude,
+            [Employee::class => $attendanceRecordOwnedEmpNumber]
+        );
+        $workflowItem = $attendanceRecord->getState() === AttendanceRecord::STATE_PUNCHED_IN ?
+            WorkflowStateMachine::ATTENDANCE_ACTION_EDIT_PUNCH_IN_TIME :
+            WorkflowStateMachine::ATTENDANCE_ACTION_EDIT_PUNCH_OUT_TIME;
+        //check whether work flow item allowed for the user
+        if (!in_array($workflowItem, array_keys($allowedWorkflowItems))) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * @throws BadRequestException
+     */
+    private function setAttendanceRecord(AttendanceRecord $attendanceRecord): AttendanceRecord
+    {
+        try {
+            list(
+                $punchInDate,
+                $punchInTime,
+                $punchInNote,
+                $punchOutDate,
+                $punchOutTime,
+                $punchOutNote
+                ) = $this->getRequestBodyParams();
+
+            $recordId = $attendanceRecord->getId();
+            $punchInTimezoneOffset = $attendanceRecord->getPunchInTimeOffset();
+            $punchInDateTime = $this->extractPunchDateTime($punchInDate.' '.$punchInTime, $punchInTimezoneOffset);
+            $punchInUTCDateTime = (clone $punchInDateTime)->setTimezone(
+                new DateTimeZone(DateTimeHelperService::TIMEZONE_UTC)
+            );
+            $attendanceRecordOwnedEmpNumber = $attendanceRecord->getEmployee()->getEmpNumber();
+            //current state is punched in and editing it
+            if ($attendanceRecord->getState() === AttendanceRecord::STATE_PUNCHED_IN) {
+                $overlappingAttendanceRecords = $this->getAttendanceService()
+                    ->getAttendanceDao()
+                    ->checkForPunchInOverLappingRecordsWhenEditing(
```

Comment:
@vageeshorangehrm shall we refactor methods names,
```suggestion
                    ->checkOverLappingForOpenRecordsWhenEditing(
```